File: .Conky/conkyx/scripts/wireless-essid.py
#!/usr/bin/env python3
'''
 Uses iwgetid to get the wireless ESSID
 Uses iwconfig to get the signal level

 When a wireless signal is rated in dBm, it is a measurement of the signal strength or power level in
 decibels (dB) referenced to one milliwatt (mW). The higher the dBm value, the stronger the signal.

 A good signal strength varies depending on the type of wireless network, the distance between the device
 and the access point, and any obstacles that may interfere with the signal. However, generally, a signal
 strength of -60 dBm to -70 dBm is considered good for a Wi-Fi network, while a cellular signal strength
 of -50 dBm to -110 dBm is considered acceptable, depending on the carrier and the location.

 It's important to note that dBm is a logarithmic scale, which means that every 3 dB increase represents
 a doubling of power. For example, a signal with a strength of -60 dBm is twice as strong as a signal with
 a strength of -63 dBm.
'''
import subprocess
import re

def get_wireless_essid(interface):
    # Use subprocess module to run the iwgetid command and get the ESSID of the wireless network
    try:
        essid = subprocess.check_output(["iwgetid", interface, "-r"], text=True).strip()
    except subprocess.CalledProcessError:
        essid = "Not connected" # If the interface is not connected to any network, set the ESSID to "Not connected"
    return essid

def get_signal_strength(interface):
    # Use subprocess module to run the iwconfig command and get the signal strength of the wireless network
    try:
        output = subprocess.check_output(["iwconfig", interface], text=True)
        # Use regular expression to search for the signal strength value in the output of iwconfig command
        match = re.search(r'Signal level=(-?\d+)', output)
        if match:
            signal_strength = int(match.group(1)) # Convert the signal strength value to an integer
            return f"{signal_strength} dBm" # Return the signal strength value in dBm unit
        else:
            return "N/A" # If signal strength value is not found in the output, set it to "N/A"
    except subprocess.CalledProcessError:
        return "N/A" # If the iwconfig command fails to run, set the signal strength to "N/A"

if __name__ == "__main__":
    interface = "wlo1" # Set the name of the wireless interface to check
    essid = get_wireless_essid(interface) # Call the get_wireless_essid function to get the ESSID of the network
    signal_strength = get_signal_strength(interface) # Call the get_signal_strength function to get the signal strength of the network
    print(f"{essid} ({signal_strength})") # Print the ESSID and signal strength in dBm unit

File: .config/.Conky/conkyx/scripts/C/Source/wireless-essid.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_BUFFER_SIZE 256

// Function to get the ESSID of the wireless network
void get_wireless_essid(const char *interface, char *essid) {
    FILE *pipe;
    char command[MAX_BUFFER_SIZE];

    snprintf(command, sizeof(command), "iwgetid %s -r", interface);
    pipe = popen(command, "r");
    if (pipe) {
        if (fgets(essid, MAX_BUFFER_SIZE, pipe) != NULL) {
            // Remove the newline character from the ESSID string
            essid[strcspn(essid, "\n")] = '\0';
        }
        pclose(pipe);
    }
}

// Function to get the signal strength of the wireless network
void get_signal_strength(const char *interface, char *signal_strength) {
    FILE *pipe;
    char command[MAX_BUFFER_SIZE];
    char output[MAX_BUFFER_SIZE];

    snprintf(command, sizeof(command), "iw dev %s link | grep signal", interface);
    pipe = popen(command, "r");
    if (pipe) {
        if (fgets(output, sizeof(output), pipe) != NULL) {
            // Extract the signal strength value from the output
            int signal_value;
            if (sscanf(output, "        signal: %d dBm", &signal_value) == 1) {
                snprintf(signal_strength, MAX_BUFFER_SIZE, "%d dBm", signal_value);
            } else {
                strncpy(signal_strength, "N/A", MAX_BUFFER_SIZE);
            }
        } else {
            strncpy(signal_strength, "N/A", MAX_BUFFER_SIZE);
        }
        pclose(pipe);
    }
}

int main() {
    const char *interface = "wlo1"; // Set the name of the wireless interface to check
    char essid[MAX_BUFFER_SIZE];
    char signal_strength[MAX_BUFFER_SIZE];

    // Call the get_wireless_essid function to get the ESSID of the network
    get_wireless_essid(interface, essid);

    // Call the get_signal_strength function to get the signal strength of the network
    get_signal_strength(interface, signal_strength);

    // Print the ESSID and signal strength in dBm unit
    printf("%s (%s)\n", essid, signal_strength);

    return 0;
}
